### PR TITLE
Adds support for logging training loop metrics, models, datasets to M…

### PR DESCRIPTION
**Pull request by narayanacharya6**
This PR is to add support for logging metrics, models and datasets to [MLflow](https://www.mlflow.org/) during spaCy training loop.

## Description
Users who wish to use the new `spacy.MLflowLogger.v1` need to have mlflow installed in their environment. The changes were tested against `mlflow==1.15.0`. Steps to replicate usage include:

1. Install `mlflow`:
`pip install mlflow`

2. Start `mlflow` tracking server locally (or skip if you already have an `mlflo` server ready and running).
`mlflow server --backend-store-uri sqlite:///mlflow.db --host localhost --port 5001 --default-artifact-root ./artifacts`

3. Setup required environment variable (change`localhost:5001` based on your setup).
`export MLFLOW_TRACKING_URI=http://localhost:5001`

4. Update `training.logger` config (below is a bare minimum config, there are other optional parameters with reasonable defaults and these are documented):
```
[training.logger]
@loggers = "spacy.MLflowLogger.v1"
experiment_name = "test_re_ner_pipe"
```

### Types of change
The changes include optional enhancement for users and also updates documentation informing how to use the enhancement. The new logger (similar to the existing WandbLogger) also outputs to the console.

## Checklist
- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy/training/loggers.py [hunk 1]
```diff
@@ -175,3 +176,68 @@ def finalize() -> None:
...
+
+                if log_losses and losses:
+                    log_metrics({f"loss_{k}": v for k, v in losses.items()}, step=step)
+
+                if log_other_scores and isinstance(other_scores, dict):
+                    log_dict(other_scores, artifact_file='scores_{}.json'.format(step))
+
+                if model_artifact_path:
+                    from spacy import load
+                    model_path = info["output_path"]
```
Comment (narayanacharya6) [inline spacy/training/loggers.py:230]: To load a saved model here just log it MLflow does not seem ideal. I'd like to have access to the `nlp` object here directly. Need input if saving the model to MLflow can be executed better if done with hooking a custom callback for `training.before_to_disk`. Any thoughts?

**Comment by narayanacharya6** [comment]
Hi @svlandeg, sorry to @ you, but I was hoping someone could review this PR, and provide some feedback :)

**Comment by svlandeg** [comment]
Hi! Thanks for the PR, and apologies for the delay in our response. There's no need to ping anyone specific though, this will certainly be reviewed at some point. There's a bit of a backlog with the PRs right now, and some are more "easy" to merge than others. I hope you understand ;-)

File: spacy/training/loggers.py (hunk 1 above)
Comment (svlandeg) [reply spacy/training/loggers.py:230]: I'm not sure I understand the issue - `setup_logger` receives the `nlp` object, right?

File: spacy/training/loggers.py [hunk 2]
```diff
@@ -175,3 +176,68 @@ def finalize() -> None:
...
+                if log_scores:
+                    log_metric("score", score, step=step)
+
+                if log_losses and losses:
+                    log_metrics({f"loss_{k}": v for k, v in losses.items()}, step=step)
+
+                if log_other_scores and isinstance(other_scores, dict):
+                    log_dict(other_scores, artifact_file='scores_{}.json'.format(step))
+
+                if model_artifact_path:
```
Comment (svlandeg) [inline spacy/training/loggers.py:228]:
```suggestion
                if model_artifact_path and info.get("output_path"):
```
We need this additional check, because `spacy train` can be run without `output_dir`.

**Comment by svlandeg** [comment]
Hi @narayanacharya6! While reviewing this PR, I noticed that `MLFlow` already has [spaCy-specific logging](https://github.com/mlflow/mlflow/blob/2cf363c43618dddacd2cef19a26c704b37561775/mlflow/spacy.py) functionality that you're calling into.

As such, we think it would make sense for this custom logger to be implemented by MLFlow instead. Then underlying changes to this logging functionality can easily be kept in-sync in the `mlflow` lib itself. So I would consider submitting this PR there.

Basically with [entry points](https://spacy.io/usage/saving-loading#entry-points), the functionality would then become available to anyone who has `mlflow` installed - it would work pretty transparently and it's probably a better solution overall. A simple example of using such an entry point can be found in our own `ml-datasets` lib, which doesn't import `spacy` but still provides custom `readers`: https://github.com/explosion/ml-datasets/blob/master/setup.cfg#L22

File: spacy/training/loggers.py (hunk 1 above)
Comment (narayanacharya6) [reply spacy/training/loggers.py:230]: Ah, missed the obvious. I was distracted by the `log_step` receiving only the `info` dictionary. My bad.

**Comment by narayanacharya6** [comment]
Thanks for reviewing the PR @svlandeg. Now that I think of it your suggestions make sense. This should reside in MLflow or a separate utility library and not within spaCy.  

Thanks for pointing me to the entry points functionality. It will be helpful for this as well as some other projects we are working on. Skimming through the documentation, it looks like this should work well for installing pipeline components (with the `spacy_factories` entry point). Will it also work for this case where we need to register a custom logger from another package and do we use the same `spacy_factories` entry point for all such custom stuff?

File: spacy/training/loggers.py (hunk 1 above)
Comment (svlandeg) [reply spacy/training/loggers.py:230]: No worries! The nested functions can sometimes be deceiving ;-)

**Comment by svlandeg** [comment]
> Will it also work for this case where we need to register a custom logger from another package and do we use the same `spacy_factories` entry point for all such custom stuff?

Ah, no. You can get an overview of the available registries in `spacy` here: https://spacy.io/api/top-level#registry. For instance, the example I linked in `ml-datasets` uses `spacy_readers` - but for this you'll want to use `spacy_loggers`.

If you'll make the PR on the `mlflow`, is it OK if I close this? I hope you can use my review comments to finalize the functionality, it looks good otherwise, and it'll be great if MLFlow can include this out-of-the box for spaCy users :-)

**Comment by narayanacharya6** [comment]
Understood, and thanks for the review comments! Yes, please feel free to
close this PR :)
